packages/metadataform-extension/src/index.ts

```diff
@@ -0,0 +1,319 @@
+          schema.metadataOptions
+        )) {
+          // Optionally links key to cell type.
+          if (options.cellTypes) {
+            if (!metaInformation[metadataKey])
+              metaInformation[metadataKey] = {};
+            metaInformation[metadataKey].cellTypes = options.cellTypes;
+          }
+
+          // Optionally links key to metadata level.
+          if (options.metadataLevel) {
+            if (!metaInformation[metadataKey])
+              metaInformation[metadataKey] = {};
+            metaInformation[metadataKey].level = options.metadataLevel;
+          }
+
+          // Optionally set the writeDefault flag.
+          if (options.writeDefault !== undefined) {
+            if (!metaInformation[metadataKey])
+              metaInformation[metadataKey] = {};
+            metaInformation[metadataKey].writeDefault = options.writeDefault;
+          }
+
+          // Optionally links key to a custom widget.
+          if (options.customWidget) {
+            const formWidget = formWidgetsRegistry.getRenderer(
+              options.customWidget as string
+            );
+
+            // If renderer is defined (custom widget has been registered), set it as used widget.
+            if (formWidget !== undefined)
+              if (!uiSchema[metadataKey]) uiSchema[metadataKey] = {};
+            uiSchema[metadataKey]['ui:widget'] = formWidget;
+          }
+
+          // Optionally links key to a custom field.
+          if (options.customField) {
+            const formField = formComponentRegistry.getRenderer(
+              options.customField as string
+            );
+
+            // If renderer is defined (custom widget has been registered), set it as used widget.
+            if (formField !== undefined)
+              if (!uiSchema[metadataKey]) uiSchema[metadataKey] = {};
+            uiSchema[metadataKey]['ui:field'] = formField;
+          }
+        }
+      }
+
+      // Adds a section to notebookTools.
+      notebookTools.addSection({
+        sectionName: schema.id,
+        rank: schema.rank,
+        label: schema.label ?? schema.id
+      });
+
+      // Creates the tool.
+      const tool = new MetadataFormWidget(
+        metadataSchema,
+        metaInformation,
+        uiSchema,
+        schema._origin,
+        translator,
+        schema.showModified
+      );
+
+      // Adds the form to the section.
+      notebookTools.addItem({ section: schema.id, tool: tool });
+
+      tools.push(tool);
+      metadataForms[schema.id] = tool;
+    }
+    return metadataForms;
+  }
+}
+
+/**
+ * The metadata form plugin.
+ */
+const metadataForm: JupyterFrontEndPlugin<
+  { [section: string]: MetadataFormWidget } | undefined
+> = {
+  id: PLUGIN_ID,
+  autoStart: true,
+  requires: [
+    INotebookTools,
+    ITranslator,
+    IFormWidgetRegistry,
+    IFormComponentRegistry
+  ],
+  optional: [ISettingRegistry],
+  provides: IMetadataFormProvider,
```

Comment:
I don't understand the need for the `IMetadataFormProvider` it is not consume by any other plugins.

Reply:
The token is necessary to update schema after the application started.

For example if we aim to rebuild the *common tools* and *advanced tools* using the *metadataform*, the `RawNBConvert` would need it. The full list of export formats is known after the application started:
https://github.com/jupyterlab/jupyterlab/blob/26999c5df85dd8a088580b5dafbc5404ef9b4a44/packages/notebook-extension/src/index.ts#L934-L943

Reply:
Thanks @brichet

In that case, could you create an interface for the provided object and encapsulate the map within an class that will have only allow to add and get the items? A good example for an simple interface like that is `IRunningSessionManagers`:

https://github.com/jupyterlab/jupyterlab/blob/219200ed47ff84c3f98355f500ac92bb48792338/packages/running/src/index.tsx#L81

Reply:
The relevant changes are included in https://github.com/jupyterlab/jupyterlab/pull/13056/commits/3f61a754ad43f82e869b7290c9d3e1603c08470e